plat/kvm/arm/smp.c

```diff
@@ -0,0 +1,239 @@
+		if (cpu_prop == NULL || prop_len <= 0) {
+			uk_pr_err("Error when reading AP reg field (cpu id) in fdt!\n");
+			continue;
+		}
+
+		id_reg = (const fdt32_t *) cpu_prop->data;
+		cpu_id = fdt_reg_read_number(id_reg, naddr);
+		if (cpu_id != bsp_cpu_id) {
+			cpu_idx++;
+			cpus[cpu_idx].id = cpu_id;
+			ukarch_spin_init(&cpus[cpu_idx].lock);
+			uk_pr_info("Initialized core (%ld)\n", cpus[cpu_idx].id);
+		}
+	}
+	smp_numcores = cpu_idx + 1;
+
+	return 0;
+}
+
+static int psci_cpu_on(struct arm64_cpu *cpu, __uptr entry)
+{
+	/* PSCI v0.1 and v0.2 both support cpu_on. the parameter @cpu after
+	 * @entry will be passed as x0 for entry
+	 */
+	return smcc_psci_hvc_call(PSCI_FNID_CPU_ON, (uint64_t) cpu->id,
+					(uint64_t) entry, (uint64_t) cpu);
+}
+
+
+int ukplat_lcpu_start(__lcpuid lcpuid[], void *sp[],
+				ukplat_lcpu_entry_t entry[], unsigned int num)
+{
+	__uptr lcpu_start_addr = ukplat_virt_to_phys(_lcpu_start);
+	uint64_t this_cpu_id = ukplat_lcpu_id();
+	__lcpuid this_cpu_lid;
+	unsigned int i, n = 0;
+	int err;
+
+	UK_ASSERT(sp);
+
+	if (num == 0) {
+		UK_ASSERT(!lcpuid);
+		num = smp_numcores;
+	}
+
+	/* start the APs */
+	for (i = 0; i < num; i++) {
+		this_cpu_lid = (lcpuid) ? lcpuid[i] : (__lcpuid) i;
+
+		if (cpus[this_cpu_lid].id == this_cpu_id)
+			continue;
+
+		if (this_cpu_lid >= smp_numcores)
+			continue;
+
+		if (cpus[this_cpu_lid].state != LCPU_STATE_OFFLINE)
+			uk_pr_warn("Core %d already starting, moving on.\n", this_cpu_lid);
+
+		if (ukarch_spin_trylock(&cpus[this_cpu_lid].lock)) {
+			cpus[this_cpu_lid].entry = entry[i] ? entry[i] : _lcpu_entry_default;
+			cpus[this_cpu_lid].stackp = sp[i];
+			cpus[this_cpu_lid].state = LCPU_STATE_STARTING;
+			UK_INIT_LIST_HEAD(&cpus[this_cpu_lid].fnlist);
+
+			barrier();
+
+			err = psci_cpu_on(&(cpus[this_cpu_lid]), lcpu_start_addr);
+			if (err != PSCI_RET_SUCCESS) {
+				uk_pr_err("Failed to start CPU %u (%lu): %d\n",
+					this_cpu_lid, cpus[this_cpu_lid].id, err);
+				cpus[this_cpu_lid].state = LCPU_STATE_OFFLINE;
+				ukarch_spin_unlock(&(cpus[this_cpu_lid].lock));
+				continue;
+			}
+
+			uk_pr_info("Started core (%lu) successfully\n", cpus[this_cpu_lid].id);
+			ukarch_spin_unlock(&cpus[this_cpu_lid].lock);
+			n++;
+		}
+	}
+
+	return n;
+}
+
+void __noreturn _lcpu_entry_default(void)
+{
+	struct arm64_cpu *this_cpu;
+
+	__asm ("str %x0, [%0]" : : "r" (&this_cpu));
+
+	intctrl_init();
+	ukplat_lcpu_enable_irq();
+
+	/* Enter halt loop. Besides interrupts in general, the halt can be
+	 * interrupted by calls to ukplat_lcpu_run().
+	 */
+	while (1) {
+		this_cpu->state = LCPU_STATE_IDLE;
+
+		/* TODO: In the interrupt handler which executes the functions
+		 * enqueued with ukplat_lcpu_run(), state should be set to
+		 * busy(3)
+		 */
+		halt();
+	}
+}
+
+uint64_t ukplat_lcpu_id(void)
+{
+	uint64_t mpidr_reg;
+
+	mpidr_reg = SYSREG_READ64(mpidr_el1);
+
+	/* return the affinity bits for the current core */
+	return mpidr_reg & 0xff00fffffful;
```

Comment:
I've changed `__u32` to `__u64` for this function. However, I think `__lcpuid` can be kept `__u32`, since it represents an index in the `cpus[]` array, not a physical id (se an earlier comment).

Reply:
Wait, are you saying that this function shall return an index? Then why does it return a part of MPIDR? If it is supposed to return an index, then it shall iterate `cpus[]` and return the index where `cpus[i].id` matches `MPIDR`. But then that will break all functions in smp.c that call this function and expect it to return an actual core id.

Also the previous comment where a 40-bit mask is applied to a 32-bit value applies still.

Reply:
No, this one returns the physical id of the current core. I have also changed the returned data type to 64-bit. The logical index will be returned by `ukplat_lcpu_idx()` (we can remove the `static` keyword there and make it part of the common API if it proves helpful for the user).